File: Daily/PY/Leetcode437-路径总和.py
class Solution:
    def pathSum(self, root: TreeNode, targetSum: int) -> int:
        if not root: return 0
        def dfs(root,targetSum):
            if not root: return 0
            if root.left == None and root.right == None:
                return 1 if root.val == targetSum else 0
            cnt = 1 if targetSum == root.val else 0
            return dfs(root.left,targetSum-root.val) + dfs(root.right,targetSum - root.val) + cnt

        return self.pathSum(root.left,targetSum) + self.pathSum(root.right,targetSum) + dfs(root,targetSum)

File: Leetcode/sbgo/Leetcode437-路径总和.go
package sbgo

func pathSum(root *TreeNode, targetSum int) int {
	if root == nil {
		return 0
	}
	return pathSum(root.Left, targetSum) + pathSum(root.Right, targetSum) + dfs(root, targetSum)
}

func dfs(root *TreeNode, targetSum int) int {
	if root == nil {
		return 0
	}
	if root.Left == nil && root.Right == nil {
		if root.Val == targetSum {
			return 1
		}
		return 0
	}
	cnt := 0
	if root.Val == targetSum {
		cnt = 1
	}
	return dfs(root.Left, targetSum-root.Val) + dfs(root.Right, targetSum-root.Val) + cnt
}
